File: Eunleelee/2024-02/2월 27일/숫자 카드 나누기.py
from math import gcd
from functools import reduce


def solution(a, b):
    a, b = list(set(a)), list(set(b))

    findGCD = lambda array: reduce(lambda acc, cur: gcd(acc, cur), array, 0)
    checkGCD = lambda array, GCD: GCD if all(e % GCD != 0 for e in array) else 0

    gcdA, gcdB = checkGCD(b, findGCD(a)), checkGCD(a, findGCD(b))

    return 0 if not (gcdA or gcdB) else max(gcdA, gcdB)

File: yht0827/2월/2월 27일/숫자 카드 나누기.java
import java.util.Arrays;

class Solution {
	public int solution(int[] arrayA, int[] arrayB) {
		int answer = 0;

		Arrays.sort(arrayA);
		Arrays.sort(arrayB);

		int gcdA = arrayA[0];
		int gcdB = arrayB[0];

		for (int i = 1; i < arrayA.length; i++) {
			gcdA = gcd(arrayA[i], gcdA);
			gcdB = gcd(arrayB[i], gcdB);
		}

		if (!isDivided(gcdA, arrayB)) {
			answer = Math.max(answer, gcdA);
		}

		if (!isDivided(gcdB, arrayA)) {
			answer = Math.max(answer, gcdB);
		}

		return answer;
	}

	public boolean isDivided(int gcd, int[] array) {
		for (int arr : array) {
			if (arr % gcd == 0)
				return true;
		}

		return false;
	}

	public int gcd(int a, int b) {
		if (b == 0)
			return a;
		else
			return gcd(b, a % b);
	}
}